File: BinarySearch/CPP/kth_missing_positive_number.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int findKthPositive(vector<int>& arr, int k) {
        int cur = 1, idx = 0, n = arr.size();
        while (true) {
            if (cur != arr[idx]) k--;
            if (cur == arr[idx] && idx < n-1) idx++;
            if (k == 0) break;
            cur++;
        }
        return cur;
    }

    int findKthPositive(vector<int>& arr, int k) {
        int l = 0, r = arr.size();
        while (l < r) {
            int m = l + (r - l) / 2;
            if (arr[m] - m - 1 < k) l = m + 1;
            else r = m;
        }
        return l + k;
    }
};

File: BinarySearch/kth_missing_positive_number.py
class Solution:
    def findKthPositive(self, arr: list[int], k: int) -> int:
        cur, idx, n = 1, 0, len(arr)
        while True:
            if cur != arr[idx]: k -= 1
            if cur == arr[idx] and idx < n-1: idx += 1
            if k == 0: break
            cur += 1
        return cur
    
    # binary search
    def findKthPositive_bs(self, arr: list[int], k: int) -> int:
        l, r = 0, len(arr)
        while l < r:
            m = (l + r) // 2
            if arr[m] - m - 1 < k: l = m + 1
            else: r = m
        return l + k